transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -155,11 +213,32 @@ impl SignatureRctBulletproofs {
             decompressed_pseudo_output_commitments.push(commitment);
         }
 
-        // Compute the pedersen generators for this token_id
-        let generator = generators(token_id);
+        // Collect list of of unique token ids
+        let token_ids = {
+            let mut token_ids = BTreeSet::default();
+            token_ids.insert(fee.token_id);
+            for token_id in &self.output_token_ids {
+                token_ids.insert(token_id.into());
+            }
+            for token_id in &self.pseudo_output_token_ids {
+                token_ids.insert(token_id.into());
```

Comment:
Confirming my understanding of transaction data: 

`pseudo_output_commitments` length is 1:1 with `inputs`, and the `token_id` set for `pseudo_output_commitments` needs to match `inputs` exactly, or else they will not properly construct commitments to zero (this is verified below when checking `Ristretto::identity` equality).

Therefore, `pseudo_output_token_ids` should have the same set of `token_ids` as `output_token_ids`, because there is no way (yet) to introduce a new `token_id` without spending an `input` from that `token_id`.

Is that correct?

Reply:
(Note: if we did things Koe's way, so that token ids are secret even from the enclave, then there would be no way to prevent this (creating zero valued outputs for nonexisting token ids). So I'm not sure it's actually worth trying to prevent it.

I guess the only reason to try to prevent it is, if people start using e.g. u64::MAX, u64::MAX - 1, etc. as token ids, then we are on the hook for actually computing those generators, which costs elliptic curve operations in prod. It would be nicer if we could say "only the first N token ids are actually enabled right now, and we cache those generators and never compute generators dynamically" as a perf optimization.

Reply:
I guess the safest thing to do is make `consensus-enclave-impl` also check these extra token ids and prohibit them if they aren't known. If we decide we don't care about this, we can remove that check later. There's no legitimate use-case that could be broken by this I think.

Reply:
This happens implicitly, since for a token have TxOuts at all, it needs to be mintable - which would imply it is configured in `tokens.toml`.

Reply:
I think that's not actually right at this revision --

I think if I have a mixed transaction, and token id = u64::MAX, with no existing TxOut's of that token id on the chain, it would work for me to add an output with zero value and this token id to the transaction builder, to any other normal transaction, and the transaction would be balanced and well-formed (since the value of that TxOut is zero). So it's not exactly minting but the TxOut would exist in the chain, and could be added as an input (even though it has zero value).